### PR TITLE
Enumerable Methods

### PR DESCRIPTION
**Pull request by macnick**
### My versions of some common Ruby enumerable methods. 

Please review my code.

Thank you

### CODE REVIEW

File: enumerable_methods.rb [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      end
+
+      def my_none?
+        if block_given?
+          !my_all? { |e| return false if yield e }
+        else
+          return !my_any? {|e| e }
+        end
+        true
+      end
```
Comment (maelfosso) [inline enumerable_methods.rb]: Please, there is a bug
```ruby
irb(main):033:0> %w{ant bear cat}.none? { |word| word.length >= 4 }
=> false
irb(main):034:0> %w{ant bear cat}.my_none? { |word| word.length >= 4 }
=> true
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello @macnick ,

# You did a really good job, just one bug to fix, only one bug

Take a look below, the function **my_none?**.
Kindly make the changes and submit a new code review request.
Thanks.

**Comment by maelfosso** [comment]
**`REQUIRED`** 
Please @macnick ,
Correctly setup *stickler-ci* and fix all the errors it will mentioned.

Thanks

**Comment by macnick** [comment]
Hello @maelfosso,

Thank you for your review. I corrected the bug in my_none? 

Also found the problem with stickler. The file was named .sticke.yml instead of .stickler.yml

Please review my code. 

Thank you.

**Comment by Torres-ssf** [comment]
### Pull Request Under Review By Sérgio (TSE)

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Hi there @macnick :hand: 
Great work implementing these changes :clap:
### **But there are still changes to be made.** We need to set up the proper linter and fix errors if it found any.

The used linter in the .stickler.yml file is not the right one. Stylelint is a CSS linter. 
https://github.com/macnick/enumerable-methods/blob/5ff9b9ddc3befd8484b890f974a7602c480849c2/.stickler.yml#L1-L3
For ruby projects, you must use Rubocop, a ruby linter:
Screenshot from 2019-09-16 07-52-39

- [ ] You will need to fix your .stickler.yml file and also the .rubocop.yml which is incomplete.

**Please check out [this](https://github.com/microverseinc/linters-config/tree/master/ruby) instruction on how to set up rubocop properly.**

Kindly make this change and submit for code review again. Happy coding mate :computer:

**Comment by macnick** [comment]
Hello @Torres-ssf,

Thank you for the helpful review. 
I did correct the stickler and rubocop files as you suggested.

Also fixed all the stickler errors except one the

enumerable_methods.rb#L92
C: Metrics/CyclomaticComplexity: Cyclomatic complexity for my_inject is too high. [9/6]
C: Metrics/PerceivedComplexity: Perceived complexity for my_inject is too high. [9/7] 

If I change this my method will not work as I want it to. 

Any suggestions?

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi team, you guys have done very well but  there are still some corrections to be made :cry: 

Please look into my_none?, my_all, my_any
```p [10, 18, 4, 6, 14, 16, 5, 8].none? { |e| e < 0 }```
 ```p [10, 18, 4, 6, 14, 16, 5, 8].my_none? { |e| e < 0 }```
The  first outputs true
The second outputs false

More details
 These are the automated test results that was failed by your program
 

- my_each_with_index returns an enumerator if no block is given
- my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
- my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
 - my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- my_any? returns true if the block ever returns a value other than false or nil 
 - my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
 - my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class
- my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
- my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- my_none? when no block or argument is given returns true only if none of the collection members is true
- my_none? when a class is passed as an argument returns true if none of the collection is a member of such class
- my_none? when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
 

With regards to the errors on stickler please see these two links [perceivedcomplxity](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/PerceivedComplexity) and [cyclomaticcomplexity](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Metrics/CyclomaticComplexity)

Please fix these things and request a review :pray:

**Comment by macnick** [comment]
Hello @Oghenebrume50,

Thank you for the review. I did all the suggested changes please 🙏 review my code again.

File: enumerable_methods.rb [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+# %w(cat dog wombat).each_with_index { |item, index|
+#   hash[item] = index
+# }
+# p hash
+# %w(cat dog wombat).my_each_with_index { |item, index|
+#   hash[item] = index
+# }
+# p hash
+# p %w(cat dog wombat).each_with_index
+# p %w(cat dog wombat).my_each_with_index
```
Comment (bolah2009) [inline enumerable_methods.rb]: Please let's delete commented out codes.

File: enumerable_methods.rb [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  #     block = stv.to_proc
+  #     stv = nil
+  #   end
+  #   sym = sym.to_sym if sym.is_a?(String)
+  #   block = sym.to_proc if sym.is_a?(Symbol)
+
+  #   # Ready to rock & roll
+  #   each { |x| stv = stv.nil? ? x : block.yield(stv, x) }
+  #   stv
+  # end
```
Comment (bolah2009) [inline enumerable_methods.rb]: #### `REQUIRED`

Consider deleting commented out codes. If you need to review a previous version of your code, use git history.

File: enumerable_methods.rb [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    if block_given?
+      (0...length).each do |i|
+        yield self[i]
+      end
+    else
+      to_enum(:my_each)
+    end
+  end
+
+  def my_each_with_index
```
Comment (bolah2009) [inline enumerable_methods.rb:14]: #### REQUIRED

There is a :bug: with `my_each_with_index` method. Please, let's fix this.

```ruby
irb(main):004:0> [1,2,3].each_with_index{|i,j| p [i, j]}
[1, 0]
[2, 1]
[3, 2]
=> [1, 2, 3]
irb(main):005:0> [1,2,3].my_each_with_index{|i,j| p [i, j]}
[1, 0]
[1, 0]
[1, 0]
=> 0...3
```

**Review by bolah2009 — CHANGES_REQUESTED** [review]
## Nice job @macnick :smile: 

#### Awesome progress... :muscle: All methods are working just fine with only one exception. :clap: 

Please see comment below, fix it and submit another code review request. 

- Update PR Description
   update-PR-description


#### :recycle:  Make these changes and submit another code review request. :+1: 
#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

File: enumerable_methods.rb (hunk 2 above)
Comment (macnick) [reply enumerable_methods.rb]: removed

File: enumerable_methods.rb (hunk 3 above)
Comment (macnick) [reply enumerable_methods.rb]: Removed all commented out lines

File: enumerable_methods.rb (hunk 4 above)
Comment (macnick) [reply enumerable_methods.rb:14]: Thank you. Fixed this also.

**Comment by Torres-ssf** [comment]
### Pull Request Under Review By Sérgio (TSE)

**Review by Torres-ssf — APPROVED** [review]
### NIcely done @macnick :clap: 

You worked hard on this project and should be proud of it :tada: :+1: 
**I'm approving this project. You can merge it now**

Good luck with the next one :smile: 
Happy coding :computer:

**Pull request merged**